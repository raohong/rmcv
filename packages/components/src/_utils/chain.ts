const chain = <T extends (...args: any[]) => void>(
  ...fns: (T | undefined)[]
): ((...args: Parameters<T>) => void) => {
  return (...args) => {
    fns.forEach((item) => {
      if (item) {
        item(...args);
      }
    });
  };
};

export default chain;
