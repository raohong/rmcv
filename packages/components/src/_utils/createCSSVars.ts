import kebabCase from 'lodash/kebabCase';

const createCSSVars = (
  prefix: string,
  style: Record<string, string | undefined | null>,
) => {
  const result = Object.entries(style).reduce((map, [key, value]) => {
    if (!value) {
      return map;
    }
    // @ts-ignore
    // eslint-disable-next-line no-param-reassign
    map[`--${prefix}-${kebabCase(key)}`] = value;

    return map;
  }, {} as React.CSSProperties);

  return result;
};

export default createCSSVars;
