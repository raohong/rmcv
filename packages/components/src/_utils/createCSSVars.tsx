import React from 'react';

export const createCSSVars = (vars: Record<string, number | string>) =>
  Object.entries(vars).reduce((result, [name, value]) => {
    // @ts-ignore
    result[`--${name}`] = value;
    return result;
  }, {} as React.CSSProperties);
