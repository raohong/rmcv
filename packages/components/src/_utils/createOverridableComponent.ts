import React, { RefAttributes } from 'react';

interface OverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
> extends React.FC<
    Omit<
      Omit<React.ComponentProps<C>, 'ref'> &
        Omit<
          JSX.IntrinsicElements[DefaultElement],
          Exclude<'ref', keyof React.ComponentProps<C>>
        >,
      keyof ExcludeProps | 'component'
    >
  > {
  // JSX.IntrinsicElements
  <Tag extends keyof JSX.IntrinsicElements>(
    props: Omit<
      Omit<
        JSX.IntrinsicElements[Tag],
        Exclude<'ref', keyof React.ComponentProps<C>>
      > &
        Omit<React.ComponentProps<C>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Tag;
    },
  ): JSX.Element;

  // ClassComponent
  <Custom extends React.ComponentClass<React.ComponentProps<Custom>>>(
    props: Omit<
      Omit<React.ComponentProps<C> & React.ComponentProps<Custom>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    } & RefAttributes<InstanceType<Custom>>,
  ): JSX.Element;

  <
    Custom extends React.ForwardRefExoticComponent<
      React.ComponentProps<Custom>
    >,
  >(
    props: Omit<
      Omit<React.ComponentProps<C>, 'ref'> & React.ComponentProps<Custom>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    },
  ): JSX.Element;
}

function createOverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
>(component: C): OverridableComponent<C, DefaultElement, ExcludeProps> {
  return component as unknown as OverridableComponent<
    C,
    DefaultElement,
    ExcludeProps
  >;
}

export default createOverridableComponent;
