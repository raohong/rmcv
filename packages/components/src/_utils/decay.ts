import { useEffect, useRef, useCallback } from 'react';
import type { SpringValue } from '@react-spring/web';
import raf from 'raf';

export type DecayAnimationOptions = {
  power?: number;
  timeConst?: number;
  from?: number;
  velocity?: number;
  modifyTarget?: (dist: number, velocity: number) => number;
  restDelta?: number;
};

type DecayAnimationState = {
  finished: boolean;
  value: number;
  velocity: number;
};

const decay = ({
  power = 0.8,
  timeConst = 300,
  from = 0,
  velocity = 0,
  restDelta = 0.2,
  modifyTarget,
}: DecayAnimationOptions) => {
  let movement = velocity * power * timeConst;
  const ideal = movement + from;

  const target = modifyTarget ? modifyTarget(ideal, velocity) : ideal;
  const state: DecayAnimationState = {
    finished: false,
    value: from,
    velocity,
  };

  if (target !== ideal) {
    movement = target - from;
  }

  return {
    next: (t: number) => {
      // y = a * Math.exp(kt)

      const dx = Math.exp((1 / -timeConst) * t);
      const delta = -movement * dx;

      state.finished = Math.abs(delta) < restDelta;
      state.value = state.finished ? target : target + delta;
      state.velocity = velocity * dx;

      return state;
    },
  };
};

export const runDecayAnimation = ({
  onChange,
  ...options
}: DecayAnimationOptions & {
  onChange: (state: DecayAnimationState) => void;
}) => {
  let id: number | null = null;
  let active = true;
  const startTime = Date.now();

  const generator = decay(options);

  const cancel = () => {
    if (id !== null) {
      raf.cancel(id);
      id = null;
    }

    active = false;
  };

  const run = () => {
    id = raf(() => {
      if (active) {
        const now = Date.now();
        const state = generator.next(now - startTime);

        onChange(state);

        if (!state.finished) {
          run();
        }
      }
    });
  };

  run();

  return cancel;
};

export const useDecayAnimation = () => {
  const unmountedRef = useRef(false);
  const canceller = useRef<(() => void)[]>([]);

  const cancelAnimation = useCallback(() => {
    if (canceller.current) {
      canceller.current.forEach((item) => item());
    }
  }, []);

  const runDecay = useCallback(
    (
      spring: SpringValue<number>,
      dir: number,
      boundary: number,
      options: DecayAnimationOptions,
    ) => {
      const cb = runDecayAnimation({
        ...options,
        onChange({ velocity, value }) {
          if (unmountedRef.current) {
            return;
          }

          if (
            (dir === 1 && value >= boundary) ||
            (dir === -1 && value <= boundary)
          ) {
            cb();
            canceller.current = canceller.current.filter((item) => item !== cb);

            spring.start({
              to: boundary,
              config: {
                decay: false,
                velocity,
              },
            });
          } else {
            spring.set(value);
          }
        },
      });

      canceller.current.push(cb);
    },
    [],
  );

  useEffect(() => {
    return () => {
      cancelAnimation();
    };
  }, [cancelAnimation]);

  return { runDecay, cancelAnimation };
};
