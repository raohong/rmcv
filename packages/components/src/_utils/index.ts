export { default as getPrefixCls } from './getPrefixCls';

export { default as isBrowser } from './isBrowser';

export { default as renderPortal } from './renderPortal';

export { default as uuid } from './uuid';

export { default as createCSSVars } from './createCSSVars';

export { default as chain } from './chain';

export { default as toArray } from './toArray';
