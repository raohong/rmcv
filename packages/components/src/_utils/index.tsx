export * from './createOverridableComponent';
export * from './decay';
export * from './flatReactNode';
export * from './getPrefixCls';
export * from './htmlAttrs';
export * from './reactRender';
export * from './useXId';
