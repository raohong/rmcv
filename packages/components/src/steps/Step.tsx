import React from 'react';
import classNames from 'classnames';
import isNil from 'lodash/isNil';
import { useConfigContext } from '../config-provider';
import type { StepProps, StepStatus } from './type';

export const STEP_SYMBOL = Symbol('Step');

const Step = React.forwardRef<HTMLDivElement, StepProps>(
  ({ children, icon, className, status, ...rest }, ref) => {
    const { getPrefixCls } = useConfigContext();
    const cls = getPrefixCls('step');

    const renderIcon = () => {
      const iconCls = getPrefixCls('step-icon');

      if (React.isValidElement(icon)) {
        return React.cloneElement(icon, {
          className: classNames(icon.props.className, iconCls),
        });
      }

      return <span className={iconCls}>{icon}</span>;
    };

    const statusList: StepStatus[] = ['process', 'finish', 'wait'];

    return (
      <div
        className={classNames(
          cls,
          {
            [`${cls}-${status}`]: status && statusList.includes(status),
          },
          className,
        )}
        ref={ref}
        {...rest}
      >
        <div className={`${cls}-title`}>{children}</div>
        <div className={`${cls}-tail`} />
        <div className={`${cls}-icon-wrapper`}>
          {isNil(icon) ? <span className={`${cls}-dot`} /> : renderIcon()}
        </div>
      </div>
    );
  },
);

// @ts-ignore
Step[STEP_SYMBOL] = true;

export default Step;
