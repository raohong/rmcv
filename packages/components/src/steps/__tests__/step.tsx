import { render, screen } from '@testing-library/react';
import React from 'react';
import { getPrefixCls } from '../../_utils';
import Step from '../Step';

const testId = 'step';

test('render correctly', () => {
  const tree = render(<Step />);

  expect(tree.asFragment()).toMatchSnapshot();
});

test('render with status', () => {
  render(<Step data-testid={testId} status="finish" />);

  expect(screen.getByTestId(testId)).toHaveClass(getPrefixCls('step-finish'));
});

test('render with icon', () => {
  render(<Step data-testid={testId} status="finish" icon={<span>icon</span>} />);

  expect(
    screen.getByTestId(testId).querySelector(`.${getPrefixCls('step-icon')}`),
  ).toContainHTML('icon');
});
