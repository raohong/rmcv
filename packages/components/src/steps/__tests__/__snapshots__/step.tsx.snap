// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  <div
    class="rmcv-step rmcv-touchable"
    style="cursor: cursor;"
  >
    <div
      class="rmcv-step-title"
    />
    <div
      class="rmcv-step-tail"
    />
    <div
      class="rmcv-step-icon-wrapper"
    >
      <span
        class="rmcv-step-dot"
      />
    </div>
  </div>
</DocumentFragment>
`;
